src/main/java/uk/gov/justice/digital/writer/Writer.java

```diff
@@ -61,70 +58,16 @@ protected static void writeCdcRecords(
     ) throws DataStorageRetriesExhaustedException {
         logger.info("Applying {} CDC records to deltalake table: {}", validRecords.count(), destinationPath);
 
-        for (Row row: validRecords.collectAsList()) {
-            processRow(spark, storage, destinationPath, primaryKey, row);
-        }
+        val inserts = validRecords.filter(col(OPERATION).equalTo(Insert.getName())).orderBy(TIMESTAMP);
+        val updates = validRecords.filter(col(OPERATION).equalTo(Update.getName())).orderBy(TIMESTAMP);
+        val deletes = validRecords.filter(col(OPERATION).equalTo(Delete.getName())).orderBy(TIMESTAMP);
+
```

Comment:
Are there any corner cases we need to worry about? This one is a bit of a stretch but, say you had a table with a PK that is a meaningful business key rather than something like an auto increment or uuid. Someone deletes it and recreates it with the same PK rather than updates it for some reason. I think you could then end up only deleting the row and not recreating it if it arrived in one batch?

Reply:
I agree. They need to be kept together.

Reply:
The entire CDC batch is now written in the same merge operation.